### ISSUE TITLE
Structs with a retriever which depend on a parent value are difficult to define

### ISSUE THREAD

**Issue by KSneijders** [Divy1211/BinaryFileParser#6]
(We discussed this in DM too, so this continues on top of that) ([URL to DM](https://discord.com/channels/@me/660231853097353278/1123175797604241478))

The only work-around that has been working is the following:

This code is inside the `__init__(...):` of the child struct:

```py
def __init__(self, struct_ver: Version = Version((0,)), parent: BaseStruct = None, idx: int = -1, initialise_defaults: bool = True, **retriever_inits):
    if parent is not None:
        Retriever.set_repeat(TerrainTables.terrain_restrictions, self, parent.terrain_count)
        Retriever.set_repeat(TerrainTables.terrain_pass_graphics, self, parent.terrain_count)
    super().__init__(struct_ver, parent, idx, initialise_defaults, **retriever_inits)
```

This doesn't feel like a proper solution in my opinion.

Could a feature (callback) be added for initializing that executes between `__init__(...)` and reading the retrievers? 
It'd still need access to `instance` and `parent` variables etc.

Thanks!

**Comment by Divy1211** [comment]
Making a void Retriever at the very beginning to set repeat using `struct.parent.value` is currently the cleanest way to do this, although I admit it might not be the best, will look into potential solutions.

example of a void Retriever: https://github.com/Divy1211/BinaryFileParser/issues/5

**Comment by Divy1211** [comment]
Starting in v0.3.0, accessing parent values is no longer supported. A new API to pass props to a child is required after v0.3.0

**Comment by Divy1211** [comment]
To allow for context sensitive parsing from not just parent object data to child object data but data anywhere in the whole struct after it is read once, I am considering adding a few new combinators:

```py
# containing class:
class Spam(BaseStruct):
    num_eggs: int = Retriever(u8, on_read = lambda: [set_repeat(Spam.eggs).from_(Spam.num_eggs)])
    num_yolks: int = Retriever(u8, on_read = lambda: [set_ctx("num_yolks").from_(Spam.num_yolks)])
    eggs: list[Egg] = Retriever(Egg)

# contained class:
class Egg(BaseStruct):
    _void = Retriever(void, on_read = lambda: [set_repeat(Egg.yolks).from_ctx("num_yolks")])
    yolks: list[int] = Retriever(u8)
```

the key that is set will persist even after `Spam` is finished parsing, so it is made accessible anywhere in the parser, not just inside `Spam` itself.

**Comment by KSneijders** [comment]
@Divy1211 Ah, so if I understand correctly, this allows to save a value in a scenario under a set key which can be retrieved at any time by any other. A "scx variable" in a way. 

I like the idea, it seems very flexible. Only thing I'm wondering, let's say I add this for an array to write to such key, and I want to retrieve only value at index 1 of the array, would that be possible?

**Comment by Divy1211** [comment]
> @Divy1211 Ah, so if I understand correctly, this allows to save a value in a scenario under a set key which can be retrieved at any time by any other. A "scx variable" in a way. 
> 
> I like the idea, it seems very flexible. Only thing I'm wondering, let's say I add this for an array to write to such key, and I want to retrieve only value at index 1 of the array, would that be possible? 

yes, like any other combinator you could do sub list or property access on context values

**Comment by KSneijders** [comment]
Sounds good, I like it!
